File: DataExtractors/Python-DataExtractor/dataExtractor.py
from google.cloud import monitoring_v3
from sys import maxsize
import pandas as pd
import json
import os
from csv import DictWriter
from dateutil import parser

class DataExtractor:
    def __init__(self, project_id, inputDirectory="pages", outputDirectory="output"):
        self.PROJECT_ID = project_id
        self.inputDirectory = inputDirectory
        self.outputDirectory = outputDirectory
        self.fieldnames = ["timestamp", "adservice","cartservice","checkoutservice","currencyservice",
                           "emailservice","frontend","paymentservice","productcatalogservice",
                           "recommendationservice","shippingservice", "label"]

        # Create the input and output directories if they don't exist
        if not os.path.exists(self.inputDirectory):
            os.makedirs(self.inputDirectory)
        if not os.path.exists(self.outputDirectory):
            os.makedirs(self.outputDirectory)

    def createFiles(self, interval, aggregation, filter="", i=1, page_token=""):
        print("{:<40}".format("Creating page json files... "), end="", flush=True)
        client = monitoring_v3.MetricServiceClient()
        # Keep calling API until we read all the pages
        while i < maxsize:
            results = client.list_time_series(
                request={
                    "name": f"projects/{self.PROJECT_ID}",
                    "filter": filter,
                    "interval": interval,
                    "view": monitoring_v3.ListTimeSeriesRequest.TimeSeriesView.FULL,
                    "aggregation": aggregation,
                    # "page_size": 25_000,
                    "page_token": page_token,
                },
                timeout=300
            )
            response = monitoring_v3.ListTimeSeriesResponse.to_json(results._response)
            fileName = f"{self.inputDirectory}/page{i}.json"
            with open(fileName, "w") as outfile:
                outfile.write(response)
            page_token = results._response.next_page_token
            if not page_token:
                break
            i += 1
        print("Done.")
                
    def mergeFiles(self):
        print("{:<40}".format("Merging page json files... "), end="", flush=True)
        output_list = []
        obj = os.scandir(self.inputDirectory)
        for entry in obj:
            if ("mergedResponse" in entry.name):
                continue
            with open(f"{self.inputDirectory}/{entry.name}", "rb") as infile:
                output_list.append(json.load(infile))

        all_time_series = []
        for json_file in output_list:
            all_time_series += json_file['timeSeries']

        textfile_merged = open(f'{self.inputDirectory}/mergedResponse.json', 'w')
        json.dump({ "timeSeries": all_time_series }, textfile_merged, indent=2)
        textfile_merged.close()
        print("Done.")

    def _getField(self, rawField):
        field = ""
        for fieldname in self.fieldnames:
            if rawField.startswith(fieldname):
                field = fieldname
        return field
    
    def _getLabel(self, attackPeriods, timestamp):
        # no need to iterate if it's clearly outside the range
        if timestamp < attackPeriods[0][0] or timestamp > attackPeriods[-1][-1]:
            return 0

        for start_time, end_time in attackPeriods:
            if timestamp >= start_time and timestamp <= end_time:
                return 1
        return 0

    def convertToCSV(self, outputFileName="output.csv", attackPeriods=None):
        print("{:<40}".format("Converting to CSV file... "), end="", flush=True)
        if not outputFileName.endswith(".csv"):
            outputFileName += ".csv"

        temp = None
        with open(f"{self.inputDirectory}/mergedResponse.json", "rb") as file:
            temp = json.load(file)
        timeSeries = temp['timeSeries']

        # first making dictionary for quick search/add 
        dataDict = dict()
        for series in timeSeries:
            field = self._getField(series['resource']['labels']['pod_name'])
            if not field:
                continue
            
            for point in series['points']:
                posixTime = parser.isoparse(point['interval']['startTime']).timestamp()
                if posixTime in dataDict.keys():
                    dataDict[posixTime][field] = point['value']['doubleValue']
                else:
                    dataDict[posixTime] = { field: point['value']['doubleValue']}

        # converting to DataFrame
        sortedKeys = list(dataDict.keys())
        sortedKeys.sort()
        dataDict = {i: dataDict[i] for i in sortedKeys}
        df = pd.DataFrame.from_dict(dataDict, orient='index')
        df = df.reindex(sorted(df.columns), axis=1)     # alphabetizing columns
        df.sort_index(axis=0)                           # sorting timestamps chronologically
        df[self.fieldnames[-1]] = df.index
        df[self.fieldnames[-1]] = df[self.fieldnames[-1]].apply(lambda x: self._getLabel(attackPeriods, x))
        df.to_csv(f"{self.outputDirectory}/{outputFileName}", index_label=self.fieldnames[0])
        print("Done.")
        

File: DataExtractors/Java-DataExtractor/src/main/java/org/example/DataExtractor.java
package org.example;

import com.google.api.gax.rpc.ApiException;
import com.google.cloud.monitoring.v3.MetricServiceClient;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.stream.JsonReader;
import com.google.monitoring.v3.Aggregation;
import com.google.monitoring.v3.ListTimeSeriesRequest;
import com.google.monitoring.v3.TimeInterval;
import com.google.monitoring.v3.TimeSeries;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;

import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.io.Writer;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.time.Instant;
import java.util.*;

public class DataExtractor {
    private final String projectID;
    private final String inputDirectory;
    private final String outputDirectory;
    private static final String MERGED_FILENAME = "mergedResponse.json";
    private final String[] fieldnames = {"timestamp", "adservice","cartservice","checkoutservice","currencyservice","emailservice","frontend","paymentservice","productcatalogservice","recommendationservice","shippingservice", "label"};

    public DataExtractor(String project_id, String inputDirectory, String outputDirectory) {
        this.projectID = project_id;
        this.inputDirectory = inputDirectory;
        this.outputDirectory = outputDirectory;
    }

    public void createFiles(TimeInterval interval, Aggregation aggregation, String filter, int i, String pageToken) throws IOException, ApiException {
        System.out.printf("%-30s", "Creating page json files... ");
        MetricServiceClient client = MetricServiceClient.create();
        Path inputDirPath = Paths.get(this.inputDirectory);
        Files.createDirectories(inputDirPath);  // create the directory if it does not exist

        Gson gson = new Gson();
        while (i < Integer.MAX_VALUE) {
            ListTimeSeriesRequest.Builder requestBuilder = ListTimeSeriesRequest.newBuilder()
                    .setName("projects/" + this.projectID)
                    .setInterval(interval)
                    .setAggregation(aggregation)
                    .setFilter(filter)
                    .setPageSize(0)
                    .setPageToken(pageToken)
                    .setView(ListTimeSeriesRequest.TimeSeriesView.FULL);

            MetricServiceClient.ListTimeSeriesPagedResponse response = client.listTimeSeries(requestBuilder.build());
            // process each TimeSeries object in the response
            for (TimeSeries ts : response.iterateAll()) {
                try (Writer writer = Files.newBufferedWriter(inputDirPath.resolve("page" + i + ".json"))) {
                    gson.toJson(ts, writer);
                }
                i++;
            }
            pageToken = response.getNextPageToken();
            if (pageToken == null || pageToken.isEmpty()) {
                break;
            }else {
                System.out.println("\tpageToken(" + i + ") = " + pageToken);
            }
        }
        client.close();
        System.out.printf("%s\n", "Done.");
    }
    public void createFiles(TimeInterval interval, Aggregation aggregation, String filter) throws IOException, ApiException {
        createFiles(interval, aggregation, filter, 0, "");
    }

    public void mergeFiles() throws IOException {
        System.out.printf("%-30s", "Merging page json files... ");
        Gson gson = new Gson();

        try (Writer writer = Files.newBufferedWriter(Paths.get(this.inputDirectory, MERGED_FILENAME))) {
            writer.write("[");  // Start of JSON array

            try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(Paths.get(this.inputDirectory))) {
                boolean firstFile = true;
                for (Path p : directoryStream) {
                    if (!p.getFileName().toString().equals(MERGED_FILENAME)) {  // Skip the mergedResponse.json file
                        if (!firstFile) {
                            writer.write(",");  // Separate JSON objects with commas
                        }
                        try (Reader reader = Files.newBufferedReader(p)) {
                            JsonObject jsonObject = gson.fromJson(reader, JsonObject.class);
                            gson.toJson(jsonObject, writer);
                            firstFile = false;
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
            writer.write("]");  // End of JSON array
        }
        System.out.printf("%s\n", "Done.");
    }

    private String getField(String rawField) {
        String field = "";
        for (String fieldname : this.fieldnames) {
            if (rawField.startsWith(fieldname)) {
                field = fieldname;
            }
        }
        return field;
    }

    private int getLabel(List<String[]> attackPeriods, long timestamp) {
        if (attackPeriods == null) {
            return 0;
        }
        for (String[] periods : attackPeriods) {
            long startTime = Instant.parse(periods[0]).getEpochSecond();
            long endTime = Instant.parse(periods[1]).getEpochSecond();
            if (timestamp >= startTime && timestamp <= endTime) {
                return 1;
            }
        }
        return 0;
    }

    public void convertToCSV(String outputFileName, List<String[]> attackPeriods) throws IOException {
        System.out.printf("%-30s", "Converting to CSV file... ");
        if (!outputFileName.endsWith(".csv")) {
            outputFileName += ".csv";
        }

        Path outputDirPath = Paths.get(this.outputDirectory);
        Files.createDirectories(outputDirPath);  // create the directory if it does not exist

        // Reading the time-series data
        Map<Long, Map<String, Double>> dataDict = new TreeMap<>();
        try (JsonReader reader = new JsonReader(new FileReader(Paths.get(this.inputDirectory, MERGED_FILENAME).toFile()))) {
            Gson gson = new Gson();

            reader.beginArray();
            while (reader.hasNext()) {
                JsonObject map = gson.fromJson(reader, JsonObject.class);

                String pod_name = map
                        .getAsJsonObject("resource_")
                        .getAsJsonObject("labels_")
                        .getAsJsonObject("mapData")
                        .get("pod_name").getAsString();
                String field = this.getField(pod_name);

                JsonArray pointsArray = map.getAsJsonArray("points_");
                for (JsonElement pointElement : pointsArray) {
                    JsonObject pointObject = pointElement.getAsJsonObject();

                    String seconds = pointObject
                            .getAsJsonObject("interval_")
                            .getAsJsonObject("endTime_")
                            .get("seconds_")
                            .getAsString();

                    Double value = pointObject
                            .getAsJsonObject("value_")
                            .get("value_")
                            .getAsDouble();

                    // Add the extracted data to dataDict
                    Long key = Long.parseLong(seconds);
                    Map<String, Double> innerMap = dataDict.computeIfAbsent(key, k -> new HashMap<>());
                    innerMap.put(field, value);
                }
            }
            reader.endArray();
        }

        // Generating the CSV from time-series data
        DecimalFormat df = new DecimalFormat("0", DecimalFormatSymbols.getInstance(Locale.ENGLISH));
        df.setMaximumFractionDigits(340); // 340 = DecimalFormat.DOUBLE_FRACTION_DIGITS
        try (CSVPrinter printer = new CSVPrinter(Files.newBufferedWriter(
                Paths.get(this.outputDirectory, outputFileName)),
                CSVFormat.DEFAULT.withHeader(this.fieldnames)) // Add this line to print the headers
        ) {
            for (Map.Entry<Long, Map<String, Double>> entry : dataDict.entrySet()) {
                List<String> record = new ArrayList<>();
                for (String fieldname : this.fieldnames) {
                    if (fieldname.equals("timestamp")) {
                        record.add(String.valueOf(entry.getKey()));
                    } else {
                        Double value = entry.getValue().get(fieldname);
                        if (value != null) {
                            record.add(df.format(value));
                        }
                    }
                }
                record.add(String.valueOf(this.getLabel(attackPeriods, entry.getKey())));
                printer.printRecord(record);
            }
        }
        System.out.printf("%s\n", "Done.");
    }

    public void convertToCSV(String outputFileName) throws IOException {
        convertToCSV(outputFileName, null);
    }
}